File: client-scalaz/src/main/scala/org/openapitools/client/api/Stop.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Stop._

case class Stop (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
latitude: Option[Double],
longitude: Option[Double],
locationDescription: Option[String],
name: Option[String],
owner: Option[Account],
location: Option[Location],
notes: Option[List[Note]],
noteCount: Option[Long],
stopType: Option[StopType],
orderIndex: Option[Long],
startTime: Option[OffsetDateTime],
endTime: Option[OffsetDateTime],
estimatedTime: Option[OffsetDateTime],
estimatedArrivalTime: Option[OffsetDateTime],
estimatedDepartureTime: Option[OffsetDateTime],
actualTime: Option[OffsetDateTime],
desiredTime: Option[OffsetDateTime],
timePadding: Option[Integer],
route: Option[Route],
shipment: Option[Shipment],
progressStatus: Option[ProgressStatus],
compareEstimatedTime: Option[OffsetDateTime],
compareEstimatedArrivalTime: Option[OffsetDateTime],
compareEstimiatedDepartureTime: Option[OffsetDateTime],
polyline: Option[String],
parent: Option[Stop],
locationDisplay: Option[String],
complete: Option[Boolean],
contentName: Option[String],
locationAddress: Option[String])

object Stop {
  import DateTimeCodecs._
  sealed trait StopType
  case object PICKUP extends StopType
  case object DROPOFF extends StopType
  case object START extends StopType
  case object END extends StopType
  case object ROUTING extends StopType

  object StopType {
    def toStopType(s: String): Option[StopType] = s match {
      case "PICKUP" => Some(PICKUP)
      case "DROPOFF" => Some(DROPOFF)
      case "START" => Some(START)
      case "END" => Some(END)
      case "ROUTING" => Some(ROUTING)
      case _ => None
    }

    def fromStopType(x: StopType): String = x match {
      case PICKUP => "PICKUP"
      case DROPOFF => "DROPOFF"
      case START => "START"
      case END => "END"
      case ROUTING => "ROUTING"
    }
  }

  implicit val StopTypeEnumEncoder: EncodeJson[StopType] =
    EncodeJson[StopType](is => StringEncodeJson(StopType.fromStopType(is)))

  implicit val StopTypeEnumDecoder: DecodeJson[StopType] =
    DecodeJson.optionDecoder[StopType](n => n.string.flatMap(jStr => StopType.toStopType(jStr)), "StopType failed to de-serialize")
  sealed trait ProgressStatus
  case object PENDING extends ProgressStatus
  case object COMPLETE extends ProgressStatus
  case object INVALID extends ProgressStatus

  object ProgressStatus {
    def toProgressStatus(s: String): Option[ProgressStatus] = s match {
      case "PENDING" => Some(PENDING)
      case "COMPLETE" => Some(COMPLETE)
      case "INVALID" => Some(INVALID)
      case _ => None
    }

    def fromProgressStatus(x: ProgressStatus): String = x match {
      case PENDING => "PENDING"
      case COMPLETE => "COMPLETE"
      case INVALID => "INVALID"
    }
  }

  implicit val ProgressStatusEnumEncoder: EncodeJson[ProgressStatus] =
    EncodeJson[ProgressStatus](is => StringEncodeJson(ProgressStatus.fromProgressStatus(is)))

  implicit val ProgressStatusEnumDecoder: DecodeJson[ProgressStatus] =
    DecodeJson.optionDecoder[ProgressStatus](n => n.string.flatMap(jStr => ProgressStatus.toProgressStatus(jStr)), "ProgressStatus failed to de-serialize")

  implicit val StopCodecJson: CodecJson[Stop] = CodecJson.derive[Stop]
  implicit val StopDecoder: EntityDecoder[Stop] = jsonOf[Stop]
  implicit val StopEncoder: EntityEncoder[Stop] = jsonEncoderOf[Stop]
}

File: client-clojure/src/sirqul_io_t_platform/specs/stop.clj
(ns sirqul-io-t-platform.specs.stop
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.location :refer :all]
            [sirqul-io-t-platform.specs.note :refer :all]
            [sirqul-io-t-platform.specs.route :refer :all]
            [sirqul-io-t-platform.specs.shipment :refer :all]
            [sirqul-io-t-platform.specs.stop :refer :all]
            )
  (:import (java.io File)))


(def stop-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :latitude) float?
   (ds/opt :longitude) float?
   (ds/opt :locationDescription) string?
   (ds/opt :name) string?
   (ds/opt :owner) account-spec
   (ds/opt :location) location-spec
   (ds/opt :notes) (s/coll-of note-spec)
   (ds/opt :noteCount) int?
   (ds/opt :stopType) string?
   (ds/opt :orderIndex) int?
   (ds/opt :startTime) inst?
   (ds/opt :endTime) inst?
   (ds/opt :estimatedTime) inst?
   (ds/opt :estimatedArrivalTime) inst?
   (ds/opt :estimatedDepartureTime) inst?
   (ds/opt :actualTime) inst?
   (ds/opt :desiredTime) inst?
   (ds/opt :timePadding) int?
   (ds/opt :route) route-spec
   (ds/opt :shipment) shipment-spec
   (ds/opt :progressStatus) string?
   (ds/opt :compareEstimatedTime) inst?
   (ds/opt :compareEstimatedArrivalTime) inst?
   (ds/opt :compareEstimiatedDepartureTime) inst?
   (ds/opt :polyline) string?
   (ds/opt :parent) stop-spec
   (ds/opt :locationDisplay) string?
   (ds/opt :complete) boolean?
   (ds/opt :contentName) string?
   (ds/opt :locationAddress) string?
   })

(def stop-spec
  (ds/spec
    {:name ::stop
     :spec stop-data}))
